File: crates/freight-fate/src/states/driving_school.rs
//! The driving school: spoken lessons on a consequence-free practice road
//! (port of `freight_fate/states/driving_school.py`).
//!
//! The school runs the real driving engine on a disposable copy of the
//! player's profile: every lesson consequence -- wear, fuel, money, hours,
//! fatigue -- lands on the copy and dies with the lesson. One guard in
//! `GameContext::save_profile` keeps the sandbox off disk, and the practice
//! drive restores the real profile on every exit path.
//!
//! Lessons are instructor state machines that duck-typed the first-run
//! `Tutorial` in Python; here they implement the same
//! [`crate::states::driving_core::Instructor`] trait, so the driving code's
//! existing `on_engine_started` / `on_parking_brake_released` /
//! `on_gear_engaged` / `update` calls drive a lesson unchanged. Instruction
//! stays on the speech channel (screen-reader rate, comma repeat, verbosity
//! all keep working); recorded instructor flavor lines can hang off the same
//! stage transitions later.
//!
//! # Shape of the port
//!
//! Python's `SchoolDrivingState` SUBCLASSED `DrivingState` to add the sandbox
//! restore on exit. Rust has no inheritance, so it wraps one and forwards
//! every `State` hook, adding `leave_sandbox` to `exit`. A lesson that
//! finishes pops itself off the stack directly (Python's `finish_lesson`
//! reached back through `self.driving`), guarded by its own `done` flag,
//! which is exactly what the `_lesson_finished` latch bought.

use ff_core::data::world_models::{Leg, Route};
use ff_core::models::jobs::{Job, CARGO_CATALOG};
use ff_core::models::profile::Profile;
use ff_core::radio::RadioState;
use ff_core::sim::vehicle::TruckState;

use crate::app::GameContext;
use crate::discord_presence::PresenceState;
use crate::impl_state_for_menu;
use crate::states::base::{InputEvent, Menu, MenuCore, MenuItem, State};
use crate::states::driving::DrivingState;
use crate::states::driving_core::{Instructor, DRIVE_PHASE_SCHOOL};

/// Long enough that no lesson meets the end of it.
pub const PRACTICE_ROAD_MILES: f64 = 25.0;

/// Swap the live profile for a throwaway copy; idempotent.
pub fn enter_sandbox(ctx: &mut GameContext) {
    if ctx.school_sandbox {
        return;
    }
    let copy = ctx.profile.as_ref().map(|p| p.duplicate());
    ctx.school_real_profile = ctx.profile.take();
    ctx.profile = copy;
    ctx.school_sandbox = true;
}

/// Restore the real profile; safe to call from every exit path.
pub fn leave_sandbox(ctx: &mut GameContext) {
    if !ctx.school_sandbox {
        return;
    }
    ctx.profile = ctx.school_real_profile.take();
    ctx.school_sandbox = false;
}

/// A flat, empty stretch of road going nowhere, starting here.
pub fn practice_route(ctx: &GameContext) -> (Job, Route) {
    let city = ctx.world.resolve_city_key(
        &ctx.profile
            .as_ref()
            .map(|p| p.current_city.clone())
            .unwrap_or_default(),
    );
    let leg = Leg::new(
        &city,
        &city,
        PRACTICE_ROAD_MILES,
        "the practice road",
        "flat",
        Vec::new(),
    );
    let mut job = Job::new(
        &CARGO_CATALOG["general"],
        0.0,
        &city,
        "the training yard",
        &city,
        PRACTICE_ROAD_MILES,
        0.0,
        24.0,
    );
    job.origin_type = "company_terminal".to_string();
    job.destination_location = "the training yard".to_string();
    job.destination_type = "company_terminal".to_string();
    job.bobtail = true;
    (job, Route::from_legs(vec![city.clone(), city], vec![leg]))
}

/// Lesson 1: engine, air, parking brake, roll to 30, stop.
///
/// Stages: 0 start the engine, 1 air and parking brake (or first gear on a
/// manual), 2 accelerate to 30, 3 brake to a full stop.
pub struct RollingBasicsLesson {
    pub stage: i32,
    pub done: bool,
    timer: f64,
    hinted: bool,
}

impl RollingBasicsLesson {
    pub const ROLL_MPH: f64 = 30.0;
    pub const STOP_MPH: f64 = 0.5;
    pub const HINT_S: f64 = 30.0;

    pub fn new() -> Self {
        RollingBasicsLesson {
            stage: 0,
            done: false,
            timer: 0.0,
            hinted: false,
        }
    }

    fn say(&self, ctx: &mut GameContext, text: String) {
        ctx.say_with(text, crate::app::Say::queued());
    }

    fn advance(&mut self, stage: i32) {
        self.stage = stage;
        self.timer = 0.0;
        self.hinted = false;
    }
}

impl Default for RollingBasicsLesson {
    fn default() -> Self {
        Self::new()
    }
}

impl Instructor for RollingBasicsLesson {
    fn begin(&mut self, ctx: &mut GameContext) {
        let text = format!(
            "Welcome to the driving school. This is the practice road: flat, empty, and none of \
             it counts. Nothing you do here touches your career, your truck, or your money. \
             Lesson one, rolling basics. First: press {} to start the engine.",
            ctx.control_hint("engine")
        );
        self.say(ctx, text);
    }

    fn on_engine_started(&mut self, ctx: &mut GameContext) {
        if self.stage != 0 {
            return;
        }
        self.advance(1);
        let text = if ctx.settings.automatic_transmission {
            format!(
                "Engine running. Let the air pressure build. When you hear air ready, press {} \
                 to release the parking brake, then hold {} to accelerate. The transmission \
                 shifts for you.",
                ctx.control_hint("parking_brake"),
                ctx.control_hint("accelerate")
            )
        } else {
            format!(
                "Engine running. Let the air pressure build. When you hear air ready, press {} \
                 to release the parking brake, hold {}, select {} for first gear, and release \
                 the clutch.",
                ctx.control_hint("parking_brake"),
                ctx.control_hint("clutch"),
                ctx.control_hint("gear_first")
            )
        };
        self.say(ctx, text);
    }

    fn on_parking_brake_released(&mut self, ctx: &mut GameContext) {
        if self.stage != 1 {
            return;
        }
        if ctx.settings.automatic_transmission {
            self.advance(2);
            let text = format!(
                "Parking brake released. Now hold {} and take it up to thirty. I will tell you \
                 when you are there.",
                ctx.control_hint("accelerate")
            );
            self.say(ctx, text);
        } else {
            self.timer = 0.0;
            self.say(
                ctx,
                "Parking brake released. Now shift into first gear.".to_string(),
            );
        }
    }

    fn on_gear_engaged(&mut self, ctx: &mut GameContext) {
        if self.stage != 1 {
            return;
        }
        self.advance(2);
        let text = format!(
            "In gear. Now hold {} and take it up to thirty. I will tell you when you are there.",
            ctx.control_hint("accelerate")
        );
        self.say(ctx, text);
    }

    fn update(&mut self, ctx: &mut GameContext, dt: f64, truck: &TruckState) {
        if self.done {
            return;
        }
        self.timer += dt;
        if self.stage == 2 && truck.speed_mph() >= Self::ROLL_MPH {
            self.advance(3);
            let text = format!(
                "Thirty. Nicely done. Now ease off and brake gently with {} to a full stop. \
                 Smooth is the goal; your freight never wants to meet the cab.",
                ctx.control_hint("brake")
            );
            self.say(ctx, text);
        } else if self.stage == 3 && truck.speed_mph() <= Self::STOP_MPH {
            self.done = true;
            self.say(
                ctx,
                "Full stop. That is the whole rhythm of the job: build it up smooth, bring it \
                 down smooth. Lesson one complete. Returning you to the school."
                    .to_string(),
            );
            // `finish_lesson`: the practice drive comes off the stack, once.
            ctx.pop_state();
        } else if self.timer > Self::HINT_S && !self.hinted {
            self.hinted = true;
            if self.stage == 0 {
                let text = format!(
                    "Reminder: press {} to start the engine.",
                    ctx.control_hint("engine")
                );
                self.say(ctx, text);
            } else if self.stage == 1 && truck.parking_brake {
                // Same contradiction the first-run tutorial had: once the air
                // is up, "wait for air" is wrong and P is the whole reminder.
                let text = if truck.air_ready() {
                    format!(
                        "Reminder: air is ready. Press {} to release the parking brake.",
                        ctx.control_hint("parking_brake")
                    )
                } else {
                    format!(
                        "Reminder: wait for air pressure to reach one hundred psi, then press {} \
                         to release the parking brake.",
                        ctx.control_hint("parking_brake")
                    )
                };
                self.say(ctx, text);
            } else if self.stage == 2 {
                let text = format!(
                    "Reminder: hold {} until you reach thirty miles per hour.",
                    ctx.control_hint("accelerate")
                );
                self.say(ctx, text);
            }
        }
    }
}

/// Which lesson a school row starts.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LessonKind {
    RollingBasics,
}

impl LessonKind {
    fn instructor(self) -> Box<dyn Instructor> {
        match self {
            LessonKind::RollingBasics => Box::new(RollingBasicsLesson::new()),
        }
    }
}

/// `LESSONS`: name, lesson, blurb.
pub const LESSONS: [(&str, LessonKind, &str); 1] = [(
    "Lesson 1: Rolling basics",
    LessonKind::RollingBasics,
    "Start the engine, build air, release the parking brake, take the truck to thirty, and brake \
     to a smooth full stop.",
)];

/// A practice drive: the real engine, a sandbox profile, an instructor.
pub struct SchoolDrivingState {
    drive: DrivingState,
}

impl SchoolDrivingState {
    pub fn new(ctx: &mut GameContext, lesson: LessonKind) -> Self {
        let (job, route) = practice_route(ctx);
        let mut drive = DrivingState::new(ctx, job, route, None, DRIVE_PHASE_SCHOOL, None);
        // The instructor rides the first-run tutorial's hooks; the school
        // replaces whatever the base state decided about the tutorial.
        drive.tutorial = Some(lesson.instructor());
        SchoolDrivingState { drive }
    }

    pub fn drive(&self) -> &DrivingState {
        &self.drive
    }

    pub fn drive_mut(&mut self) -> &mut DrivingState {
        &mut self.drive
    }
}

impl State for SchoolDrivingState {
    fn paces_main_speech(&self) -> bool {
        self.drive.paces_main_speech()
    }

    fn enter(&mut self, ctx: &mut GameContext) {
        self.drive.enter(ctx);
    }

    fn exit(&mut self, ctx: &mut GameContext) {
        self.drive.exit(ctx);
        // Every path out of a practice drive restores the real profile:
        // lesson completion, Escape, even an abandon from the pause menu.
        leave_sandbox(ctx);
    }

    fn handle_event(&mut self, ctx: &mut GameContext, event: &InputEvent) {
        self.drive.handle_event(ctx, event);
    }

    fn handle_controller(&mut self, ctx: &mut GameContext, event: &InputEvent) {
        self.drive.handle_controller(ctx, event);
    }

    fn on_controller_disconnect(&mut self, ctx: &mut GameContext) {
        self.drive.on_controller_disconnect(ctx);
    }

    fn update(&mut self, ctx: &mut GameContext, dt: f64) {
        self.drive.update(ctx, dt);
    }

    fn lines(&self, ctx: &GameContext) -> Vec<String> {
        self.drive.lines(ctx)
    }

    fn presence(&self, ctx: &GameContext) -> Option<PresenceState> {
        self.drive.presence(ctx)
    }

    fn online_presence(&self, ctx: &GameContext) -> Option<PresenceState> {
        self.drive.online_presence(ctx)
    }

    fn ticks_covered_music(&self) -> bool {
        self.drive.ticks_covered_music()
    }

    fn tick_covered_music(&mut self, ctx: &mut GameContext, dt: f64) {
        self.drive.tick_covered_music(ctx, dt);
    }

    fn applies_radio_settings(&self) -> bool {
        self.drive.applies_radio_settings()
    }

    fn apply_radio_settings_now(&mut self, ctx: &mut GameContext) {
        self.drive.apply_radio_settings_now(ctx);
    }

    fn radio(&self) -> Option<&RadioState> {
        self.drive.radio()
    }
}

pub struct DrivingSchoolState {
    menu: MenuCore<Self>,
}

const SCHOOL_INTRO_HELP: &str =
    "Pick a lesson. Lessons run on a practice road where nothing counts: no money, no wear, no \
     hours. Escape returns to the terminal.";

impl DrivingSchoolState {
    pub fn new() -> Self {
        DrivingSchoolState {
            menu: MenuCore::new("Driving school").with_intro_help(SCHOOL_INTRO_HELP),
        }
    }

    fn start(&mut self, ctx: &mut GameContext, lesson: LessonKind) {
        enter_sandbox(ctx);
        let state = SchoolDrivingState::new(ctx, lesson);
        ctx.push_state(state);
    }
}

impl Default for DrivingSchoolState {
    fn default() -> Self {
        Self::new()
    }
}

impl Menu for DrivingSchoolState {
    fn menu(&self) -> &MenuCore<Self> {
        &self.menu
    }

    fn menu_mut(&mut self) -> &mut MenuCore<Self> {
        &mut self.menu
    }

    fn enter(&mut self, ctx: &mut GameContext) {
        // Belt and suspenders: if a practice drive ever unwinds without its
        // exit hook, re-entering the school still restores the player.
        leave_sandbox(ctx);
        let items = self.build_items(ctx);
        self.menu.items = items;
        self.menu.index = self.menu.index.min(self.menu.items.len().saturating_sub(1));
        if let Some(key) = self.menu.open_sound_key.clone() {
            ctx.audio.play(&key);
        }
        self.announce_entry(ctx);
    }

    fn build_items(&mut self, _ctx: &mut GameContext) -> Vec<MenuItem<Self>> {
        let mut items: Vec<MenuItem<Self>> = LESSONS
            .iter()
            .map(|(name, lesson, blurb)| {
                let lesson = *lesson;
                MenuItem::new(*name, move |s: &mut Self, ctx: &mut GameContext| {
                    s.start(ctx, lesson)
                })
                .help(*blurb)
            })
            .collect();
        items.push(
            MenuItem::new("Back to terminal", |s: &mut Self, ctx| s.go_back(ctx))
                .help("Leave the school."),
        );
        items
    }

    fn go_back(&mut self, ctx: &mut GameContext) {
        ctx.audio.play("ui/menu_back");
        ctx.pop_state();
    }
}

impl_state_for_menu!(DrivingSchoolState);

/// `Profile.from_dict(profile.to_dict())`: the throwaway copy the sandbox
/// runs on.
trait DuplicateProfile {
    fn duplicate(&self) -> Profile;
}

impl DuplicateProfile for Profile {
    fn duplicate(&self) -> Profile {
        Profile::from_dict(&self.to_dict())
    }
}

File: src/freight_fate/states/driving_school.py
"""The driving school: spoken lessons on a consequence-free practice road.

The school runs the real driving engine on a disposable copy of the
player's profile: every lesson consequence -- wear, fuel, money, hours,
fatigue -- lands on the copy and dies with the lesson. One guard in
``GameContext.save_profile`` keeps the sandbox off disk, and the practice
drive restores the real profile on every exit path.

Lessons are instructor state machines that duck-type the first-run
``Tutorial`` (see driving_core): the driving code already calls
``on_engine_started`` / ``on_parking_brake_released`` / ``on_gear_engaged``
/ ``update``, so a lesson plugs into those hooks unchanged. Instruction
stays on the speech channel (screen-reader rate, comma repeat, verbosity
all keep working); recorded instructor flavor lines can hang off the same
stage transitions later.
"""

from __future__ import annotations

from ..data.world_models import Leg, Route
from ..models.jobs import CARGO_CATALOG, Job
from ..models.profile import Profile
from .base import MenuItem, MenuState
from .driving import DrivingState
from .driving_core import DRIVE_PHASE_SCHOOL

PRACTICE_ROAD_MILES = 25.0  # long enough that no lesson meets the end of it


def _enter_sandbox(ctx) -> None:
    """Swap the live profile for a throwaway copy; idempotent."""
    if getattr(ctx, "school_sandbox", False):
        return
    ctx.school_real_profile = ctx.profile
    ctx.profile = Profile.from_dict(ctx.profile.to_dict())
    ctx.school_sandbox = True


def _leave_sandbox(ctx) -> None:
    """Restore the real profile; safe to call from every exit path."""
    if not getattr(ctx, "school_sandbox", False):
        return
    ctx.profile = ctx.school_real_profile
    ctx.school_real_profile = None
    ctx.school_sandbox = False


def _practice_route(ctx) -> tuple[Job, Route]:
    """A flat, empty stretch of road going nowhere, starting here."""
    p = ctx.profile
    city = ctx.world.resolve_city_key(p.current_city)
    leg = Leg(
        a=city,
        b=city,
        miles=PRACTICE_ROAD_MILES,
        highway="the practice road",
        terrain="flat",
        stops=(),
    )
    job = Job(
        CARGO_CATALOG["general"],
        0.0,
        city,
        "the training yard",
        city,
        PRACTICE_ROAD_MILES,
        0.0,
        24.0,
        origin_type="company_terminal",
        destination_location="the training yard",
        destination_type="company_terminal",
        bobtail=True,
    )
    return job, Route([city, city], [leg])


class RollingBasicsLesson:
    """Lesson 1: engine, air, parking brake, roll to 30, stop.

    Duck-types the first-run Tutorial so the existing driving hooks drive
    the lesson. Stages: 0 start the engine, 1 air and parking brake (or
    first gear on a manual), 2 accelerate to 30, 3 brake to a full stop.
    """

    ROLL_MPH = 30.0
    STOP_MPH = 0.5
    HINT_S = 30.0

    def __init__(self, ctx, driving) -> None:
        self.ctx = ctx
        self.driving = driving
        self.stage = 0
        self.done = False
        self._timer = 0.0
        self._hinted = False

    def _say(self, text: str) -> None:
        self.ctx.say(text, interrupt=False)

    def _advance(self, stage: int) -> None:
        self.stage = stage
        self._timer = 0.0
        self._hinted = False

    def begin(self) -> None:
        self._say(
            "Welcome to the driving school. This is the practice road: "
            "flat, empty, and none of it counts. Nothing you do here "
            "touches your career, your truck, or your money. "
            "Lesson one, rolling basics. First: press "
            f"{self.ctx.control_hint('engine')} to start the engine."
        )

    def on_engine_started(self) -> None:
        if self.stage != 0:
            return
        self._advance(1)
        if self.ctx.settings.automatic_transmission:
            self._say(
                "Engine running. Let the air pressure build. When you hear "
                f"air ready, press {self.ctx.control_hint('parking_brake')} "
                "to release the parking brake, then hold "
                f"{self.ctx.control_hint('accelerate')} to accelerate. "
                "The transmission shifts for you."
            )
        else:
            self._say(
                "Engine running. Let the air pressure build. When you hear "
                f"air ready, press {self.ctx.control_hint('parking_brake')} "
                "to release the parking brake, hold "
                f"{self.ctx.control_hint('clutch')}, select "
                f"{self.ctx.control_hint('gear_first')} for first gear, and "
                "release the clutch."
            )

    def on_parking_brake_released(self) -> None:
        if self.stage != 1:
            return
        if self.ctx.settings.automatic_transmission:
            self._advance(2)
            self._say(
                "Parking brake released. Now hold "
                f"{self.ctx.control_hint('accelerate')} and take it up to "
                "thirty. I will tell you when you are there."
            )
        else:
            self._timer = 0.0
            self._say("Parking brake released. Now shift into first gear.")

    def on_gear_engaged(self) -> None:
        if self.stage != 1:
            return
        self._advance(2)
        self._say(
            "In gear. Now hold "
            f"{self.ctx.control_hint('accelerate')} and take it up to "
            "thirty. I will tell you when you are there."
        )

    def update(self, dt: float, truck) -> None:
        if self.done:
            return
        self._timer += dt
        if self.stage == 2 and truck.speed_mph >= self.ROLL_MPH:
            self._advance(3)
            self._say(
                "Thirty. Nicely done. Now ease off and brake gently with "
                f"{self.ctx.control_hint('brake')} to a full stop. "
                "Smooth is the goal; your freight never wants to meet the cab."
            )
        elif self.stage == 3 and truck.speed_mph <= self.STOP_MPH:
            self.done = True
            self._say(
                "Full stop. That is the whole rhythm of the job: build it "
                "up smooth, bring it down smooth. Lesson one complete. "
                "Returning you to the school."
            )
            self.driving.finish_lesson()
        elif self._timer > self.HINT_S and not self._hinted:
            self._hinted = True
            if self.stage == 0:
                self._say(
                    f"Reminder: press {self.ctx.control_hint('engine')} to "
                    "start the engine."
                )
            elif self.stage == 1 and truck.parking_brake:
                self._say(
                    "Reminder: wait for air pressure to reach one hundred "
                    f"psi, then press {self.ctx.control_hint('parking_brake')} "
                    "to release the parking brake."
                )
            elif self.stage == 2:
                self._say(
                    "Reminder: hold "
                    f"{self.ctx.control_hint('accelerate')} until you reach "
                    "thirty miles per hour."
                )


LESSONS = (
    (
        "Lesson 1: Rolling basics",
        RollingBasicsLesson,
        "Start the engine, build air, release the parking brake, take the "
        "truck to thirty, and brake to a smooth full stop.",
    ),
)


class SchoolDrivingState(DrivingState):
    """A practice drive: the real engine, a sandbox profile, an instructor."""

    def __init__(self, ctx, lesson_cls) -> None:
        job, route = _practice_route(ctx)
        super().__init__(ctx, job, route, phase=DRIVE_PHASE_SCHOOL)
        # The instructor rides the first-run tutorial's hooks; the school
        # replaces whatever the base state decided about the tutorial.
        self.tutorial = lesson_cls(ctx, self)
        self._lesson_finished = False

    def finish_lesson(self) -> None:
        if self._lesson_finished:
            return
        self._lesson_finished = True
        self.ctx.pop_state()

    def exit(self) -> None:
        super().exit()
        # Every path out of a practice drive restores the real profile:
        # lesson completion, Escape, even an abandon from the pause menu.
        _leave_sandbox(self.ctx)


class DrivingSchoolState(MenuState):
    title = "Driving school"
    intro_help = (
        "Pick a lesson. Lessons run on a practice road where nothing "
        "counts: no money, no wear, no hours. Escape returns to the "
        "terminal."
    )

    def enter(self) -> None:
        # Belt and suspenders: if a practice drive ever unwinds without
        # its exit hook, re-entering the school still restores the player.
        _leave_sandbox(self.ctx)
        super().enter()

    def build_items(self) -> list[MenuItem]:
        items = [
            MenuItem(
                name,
                lambda cls=lesson_cls: self._start(cls),
                help=blurb,
            )
            for name, lesson_cls, blurb in LESSONS
        ]
        items.append(
            MenuItem("Back to terminal", self.go_back, help="Leave the school.")
        )
        return items

    def _start(self, lesson_cls) -> None:
        _enter_sandbox(self.ctx)
        self.ctx.push_state(SchoolDrivingState(self.ctx, lesson_cls))

    def go_back(self) -> None:
        self.ctx.audio.play("ui/menu_back")
        self.ctx.pop_state()
